File: apps/runtime/src/services/library/settings-sources.ts
/**
 * Raw settings and hook sources from the machine that holds them.
 *
 * Only the opening of files happens here. Parsing, redaction, and the concept
 * comparison across targets are hub decisions over these bytes — they need no
 * filesystem, and keeping them hub-side means one parser, not one per host.
 *
 * The location set comes from the target registry rather than a list in this
 * file, so a new settings location is a registry row and not a method change.
 * Two locations can name the same file (Claude's settings and hooks are one
 * `settings.json`), and it is opened once.
 */

import {
  closeSync,
  type Dirent,
  constants as fsConstants,
  fstatSync,
  openSync,
  readdirSync,
  readSync,
} from 'node:fs';
import { join } from 'node:path';
import type { LibraryLocationId } from '@mangostudio/shared/library';
import {
  getLibraryLocation,
  LIBRARY_TARGET_DEFINITIONS,
  type LocationDefinition,
} from '@mangostudio/shared/library/host';
import type { PathEnv } from '@mangostudio/shared/runtime-env';

/** Ceiling for one settings source, matching the pre-relocation hub reader. */
const MAX_SETTINGS_SOURCE_BYTES = 512 * 1024;

// O_NOFOLLOW makes a final-component symlink fail the open instead of silently
// resolving to its target: a settings path is a fixed, vendor-defined name, so
// anything it points at elsewhere is not the file the user was asked about.
const O_NOFOLLOW = fsConstants.O_NOFOLLOW ?? 0;
const READ_FLAGS = fsConstants.O_RDONLY | O_NOFOLLOW;

/** Why a source that exists could not be turned into settings text. */
type RuntimeSettingsReadFailure = 'unreadable' | 'not-regular-file' | 'too-large';

interface RuntimeSettingsRuleFile {
  readonly name: string;
  readonly content: string;
}

export interface RuntimeSettingsSource {
  readonly locationId: LibraryLocationId;
  /**
   * False when the location does not resolve on this machine or nothing is
   * there. A missing settings file is an ordinary state, not a failure.
   */
  readonly present: boolean;
  readonly sizeBytes?: number;
  readonly failureReason?: RuntimeSettingsReadFailure;
  /** Raw text, for every format except `rules-dsl`. */
  readonly content?: string;
  /** One entry per `.rules` file, name-sorted, for `rules-dsl` locations. */
  readonly rules?: readonly RuntimeSettingsRuleFile[];
}

export interface RuntimeSettingsSourcesResult {
  /**
   * This machine's home directory. The hub's parsers abbreviate paths against
   * it, and abbreviating a remote path against the hub's home would be wrong.
   */
  readonly homeDir: string;
  readonly sources: readonly RuntimeSettingsSource[];
}

/** Every location any target reads settings or hooks from, in registry order. */
function settingsSourceLocationIds(): LibraryLocationId[] {
  const seen = new Set<LibraryLocationId>();
  for (const target of LIBRARY_TARGET_DEFINITIONS) {
    for (const locationId of [...target.reads.setting, ...target.reads.hook]) {
      seen.add(locationId);
    }
  }
  return [...seen];
}

export function readSettingsSources(env: PathEnv): RuntimeSettingsSourcesResult {
  const byPath = new Map<string, RuntimeSettingsSource>();
  const sources = settingsSourceLocationIds().map((locationId) => {
    const location = getLibraryLocation(locationId);
    if (!location) return { locationId, present: false };
    const path = location.resolvePath(env);
    if (path === null) return { locationId, present: false };

    const cached = byPath.get(path);
    if (cached) return { ...cached, locationId };
    const source = { ...readSource(location, path), locationId };
    byPath.set(path, source);
    return source;
  });
  return { homeDir: env.homeDir, sources };
}

function readSource(location: LocationDefinition, path: string): RuntimeSettingsSource {
  const locationId = location.id;
  if (location.format === 'rules-dsl') {
    const read = readRulesDirectory(path);
    // null is "nothing there" — same as a missing settings file, not an empty
    // present directory (which would report present: true with rules: []).
    if (read === null) return { locationId, present: false };
    return read.failureReason === undefined
      ? { locationId, present: true, sizeBytes: read.sizeBytes, rules: read.rules }
      : failed(locationId, read.failureReason);
  }

  try {
    const { content, sizeBytes } = readBoundedUtf8(path);
    return { locationId, present: true, sizeBytes, content };
  } catch (error) {
    const reason = classifyReadError(error);
    return reason === null ? { locationId, present: false } : failed(locationId, reason);
  }
}

function failed(
  locationId: LibraryLocationId,
  failureReason: RuntimeSettingsReadFailure
): RuntimeSettingsSource {
  return { locationId, present: true, failureReason };
}

interface RulesDirectoryRead {
  readonly rules: readonly RuntimeSettingsRuleFile[];
  readonly sizeBytes: number;
  readonly failureReason?: RuntimeSettingsReadFailure;
}

/**
 * Null means the directory is absent (`ENOENT`). An empty existing directory
 * is a successful present read with `rules: []`.
 */
function readRulesDirectory(path: string): RulesDirectoryRead | null {
  let entries: Dirent<string>[];
  try {
    entries = readdirSync(path, { withFileTypes: true });
  } catch (error) {
    const code = (error as NodeJS.ErrnoException | undefined)?.code;
    if (code === 'ENOENT') return null;
    return {
      rules: [],
      sizeBytes: 0,
      failureReason: code === 'ENOTDIR' ? 'not-regular-file' : 'unreadable',
    };
  }

  const rules: RuntimeSettingsRuleFile[] = [];
  let sizeBytes = 0;
  for (const entry of entries.sort((left, right) =>
    left.name < right.name ? -1 : left.name > right.name ? 1 : 0
  )) {
    if (entry.name.startsWith('.') || !entry.isFile() || !entry.name.endsWith('.rules')) continue;
    let file: { content: string; sizeBytes: number };
    try {
      file = readBoundedUtf8(join(path, entry.name));
    } catch (error) {
      // A file unlinked between readdir and open must not turn a directory that
      // demonstrably exists into an absent source.
      if (classifyReadError(error) === null) continue;
      return { rules: [], sizeBytes: 0, failureReason: 'unreadable' };
    }
    sizeBytes += file.sizeBytes;
    if (sizeBytes > MAX_SETTINGS_SOURCE_BYTES) {
      return { rules: [], sizeBytes: 0, failureReason: 'too-large' };
    }
    rules.push({ name: entry.name, content: file.content });
  }
  return { rules, sizeBytes };
}

/**
 * Opens once and validates the descriptor with `fstat`, rather than stat-ing a
 * path another process could swap in between.
 */
function readBoundedUtf8(path: string): { content: string; sizeBytes: number } {
  const fd = openSync(path, READ_FLAGS);
  try {
    const stats = fstatSync(fd);
    if (!stats.isFile()) throw new SettingsReadError('not-regular-file');
    const sizeBytes = stats.size;
    if (sizeBytes > MAX_SETTINGS_SOURCE_BYTES) throw new SettingsReadError('too-large');
    if (sizeBytes === 0) return { content: '', sizeBytes };

    const buffer = Buffer.alloc(sizeBytes);
    let offset = 0;
    while (offset < sizeBytes) {
      const bytesRead = readSync(fd, buffer, offset, sizeBytes - offset, offset);
      if (bytesRead === 0) break;
      offset += bytesRead;
    }
    return { content: buffer.subarray(0, offset).toString('utf8'), sizeBytes };
  } finally {
    closeSync(fd);
  }
}

class SettingsReadError extends Error {
  constructor(readonly reason: RuntimeSettingsReadFailure) {
    super(`Cannot read settings source: ${reason}`);
    this.name = 'SettingsReadError';
  }
}

/** Null means "nothing there" — the one outcome that is not a failure. */
function classifyReadError(error: unknown): RuntimeSettingsReadFailure | null {
  if (error instanceof SettingsReadError) return error.reason;
  const code = (error as NodeJS.ErrnoException | undefined)?.code;
  if (code === 'ENOENT') return null;
  // ELOOP: O_NOFOLLOW rejected a symlink. EISDIR: a directory where supported.
  if (code === 'ELOOP' || code === 'EISDIR') return 'not-regular-file';
  return 'unreadable';
}

File: crates/mangostudio-runtime/src/library/settings_sources.rs
//! `library.settings-sources`: the raw bytes of every settings and hook
//! source on this machine — `readSettingsSources` from
//! `apps/shared/src/library/machine/settings-sources.ts`.
//!
//! Only opening happens here; parsing, redaction and comparison are the
//! hub's. A missing file is `present: false` (an ordinary state, not a
//! failure); a file that exists but cannot be turned into text reports why
//! (`unreadable`, `not-regular-file`, `too-large`). A settings path is a
//! fixed vendor name, so a final-component symlink is refused rather than
//! followed (`O_NOFOLLOW` where the platform has it, as TypeScript does).

use std::collections::HashMap;
use std::fs::OpenOptions;
use std::io::Read;
use std::path::Path;

use serde::Serialize;

use super::js::{buffer_to_utf8_string, cmp_utf16};
use crate::probing::detection::agent_cli_definitions::AgentTargetId;
use crate::probing::detection::path_env::PathEnv;
use crate::probing::locations::{
    LocationDefinition, ResourceFormat, location_by_id, location_ids_for_target,
};

/// `MAX_SETTINGS_SOURCE_BYTES`, for one file or one rules directory's total.
const MAX_SETTINGS_SOURCE_BYTES: u64 = 512 * 1024;

/// `RuntimeSettingsReadFailure`.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize)]
#[serde(rename_all = "kebab-case")]
pub(crate) enum ReadFailure {
    Unreadable,
    NotRegularFile,
    TooLarge,
}

/// `RuntimeSettingsRuleFile`.
#[derive(Debug, Clone, PartialEq, Eq, Serialize)]
pub(crate) struct RuleFile {
    pub name: String,
    pub content: String,
}

/// `RuntimeSettingsSource`.
#[derive(Debug, Clone, PartialEq, Eq, Serialize)]
#[serde(rename_all = "camelCase")]
pub(crate) struct Source {
    pub location_id: &'static str,
    pub present: bool,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub size_bytes: Option<u64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub failure_reason: Option<ReadFailure>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub content: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub rules: Option<Vec<RuleFile>>,
}

impl Source {
    fn absent(location_id: &'static str) -> Self {
        Self {
            location_id,
            present: false,
            size_bytes: None,
            failure_reason: None,
            content: None,
            rules: None,
        }
    }

    fn failed(location_id: &'static str, reason: ReadFailure) -> Self {
        Self {
            present: true,
            failure_reason: Some(reason),
            ..Self::absent(location_id)
        }
    }
}

/// `RuntimeSettingsSourcesResult`.
#[derive(Debug, Clone, PartialEq, Eq, Serialize)]
#[serde(rename_all = "camelCase")]
pub(crate) struct SettingsSources {
    pub home_dir: String,
    pub sources: Vec<Source>,
}

/// Every location a target reads settings or hooks from, in target then
/// registry order, deduplicated — `settingsSourceLocationIds`. Each
/// target's list already orders settings before hooks.
#[must_use]
pub(crate) fn settings_source_location_ids() -> Vec<&'static str> {
    let mut ids: Vec<&'static str> = Vec::new();
    for target in [
        AgentTargetId::Mangostudio,
        AgentTargetId::Claude,
        AgentTargetId::Codex,
        AgentTargetId::Cursor,
    ] {
        for id in location_ids_for_target(target) {
            let kind = location_by_id(id).map(|location| location.kind);
            if matches!(kind, Some("setting" | "hook")) && !ids.contains(id) {
                ids.push(id);
            }
        }
    }
    ids
}

/// `readSettingsSources`. Two locations naming one file read it once.
///
/// # Example
///
/// ```ignore
/// let sources = read_settings_sources(&env);
/// assert_eq!(sources.sources[0].location_id, "mango-settings");
/// ```
#[must_use]
pub(crate) fn read_settings_sources(env: &PathEnv) -> SettingsSources {
    let mut by_path: HashMap<String, Source> = HashMap::new();
    let sources = settings_source_location_ids()
        .into_iter()
        .map(|id| {
            let Some(location) = location_by_id(id) else {
                return Source::absent(id);
            };
            let Some(path) = (location.resolve_path)(env) else {
                return Source::absent(location.id);
            };
            if let Some(cached) = by_path.get(&path) {
                return Source {
                    location_id: location.id,
                    ..cached.clone()
                };
            }
            let source = read_source(location, &path);
            by_path.insert(path, source.clone());
            source
        })
        .collect();
    SettingsSources {
        home_dir: env.home_dir.clone(),
        sources,
    }
}

fn read_source(location: &LocationDefinition, path: &str) -> Source {
    if location.format == ResourceFormat::RulesDsl {
        return match read_rules_directory(Path::new(path)) {
            None => Source::absent(location.id),
            Some(Ok((rules, size_bytes))) => Source {
                present: true,
                size_bytes: Some(size_bytes),
                rules: Some(rules),
                ..Source::absent(location.id)
            },
            Some(Err(reason)) => Source::failed(location.id, reason),
        };
    }
    match read_bounded(Path::new(path)) {
        Ok((content, size_bytes)) => Source {
            present: true,
            size_bytes: Some(size_bytes),
            content: Some(content),
            ..Source::absent(location.id)
        },
        Err(None) => Source::absent(location.id),
        Err(Some(reason)) => Source::failed(location.id, reason),
    }
}

type RulesRead = Result<(Vec<RuleFile>, u64), ReadFailure>;

/// `None` is an absent directory; an empty existing one is `Ok(([], 0))`.
fn read_rules_directory(path: &Path) -> Option<RulesRead> {
    let entries = match std::fs::read_dir(path) {
        Ok(entries) => entries,
        Err(error) if error.kind() == std::io::ErrorKind::NotFound => return None,
        Err(error) if error.kind() == std::io::ErrorKind::NotADirectory => {
            return Some(Err(ReadFailure::NotRegularFile));
        }
        Err(_) => return Some(Err(ReadFailure::Unreadable)),
    };
    let mut files: Vec<(String, bool)> = Vec::new();
    for entry in entries {
        let Ok(entry) = entry else {
            return Some(Err(ReadFailure::Unreadable));
        };
        // `Dirent.isFile()` does not follow a symlink.
        let is_file = entry.file_type().is_ok_and(|kind| kind.is_file());
        files.push((entry.file_name().to_string_lossy().into_owned(), is_file));
    }
    files.sort_by(|(left, _), (right, _)| cmp_utf16(left, right));

    let mut rules = Vec::new();
    let mut size_bytes = 0u64;
    for (name, is_file) in files {
        if name.starts_with('.') || !is_file || !name.ends_with(".rules") {
            continue;
        }
        let (content, file_bytes) = match read_bounded(&path.join(&name)) {
            Ok(read) => read,
            // Unlinked between listing and open: skip it, the directory
            // itself still demonstrably exists.
            Err(None) => continue,
            Err(Some(_)) => return Some(Err(ReadFailure::Unreadable)),
        };
        size_bytes += file_bytes;
        if size_bytes > MAX_SETTINGS_SOURCE_BYTES {
            return Some(Err(ReadFailure::TooLarge));
        }
        rules.push(RuleFile { name, content });
    }
    Some(Ok((rules, size_bytes)))
}

/// `readBoundedUtf8` plus `classifyReadError`: `Err(None)` is "nothing
/// there", `Err(Some(reason))` a real failure.
fn read_bounded(path: &Path) -> Result<(String, u64), Option<ReadFailure>> {
    let file = open_no_follow(path).map_err(|error| classify(path, &error))?;
    let metadata = file.metadata().map_err(|_| Some(ReadFailure::Unreadable))?;
    if !metadata.is_file() {
        return Err(Some(ReadFailure::NotRegularFile));
    }
    if metadata.len() > MAX_SETTINGS_SOURCE_BYTES {
        return Err(Some(ReadFailure::TooLarge));
    }
    let mut bytes = Vec::with_capacity(usize::try_from(metadata.len()).unwrap_or(0));
    file.take(metadata.len())
        .read_to_end(&mut bytes)
        .map_err(|_| Some(ReadFailure::Unreadable))?;
    Ok((buffer_to_utf8_string(&bytes), metadata.len()))
}

fn open_no_follow(path: &Path) -> std::io::Result<std::fs::File> {
    let mut options = OpenOptions::new();
    options.read(true);
    #[cfg(unix)]
    {
        use std::os::unix::fs::OpenOptionsExt;
        // Non-blocking as well, so a FIFO swapped in cannot wedge the worker
        // before the `fstat` check refuses it.
        options.custom_flags(nix::libc::O_NOFOLLOW | nix::libc::O_NONBLOCK);
    }
    options.open(path)
}

fn classify(path: &Path, error: &std::io::Error) -> Option<ReadFailure> {
    if error.kind() == std::io::ErrorKind::NotFound {
        return None;
    }
    #[cfg(unix)]
    if matches!(
        error.raw_os_error(),
        Some(nix::libc::ELOOP | nix::libc::EISDIR)
    ) {
        return Some(ReadFailure::NotRegularFile);
    }
    // Windows refuses to open a directory at all (Node reports EISDIR).
    if cfg!(windows) && path.is_dir() {
        return Some(ReadFailure::NotRegularFile);
    }
    Some(ReadFailure::Unreadable)
}

#[cfg(test)]
mod tests {
    use super::*;

    fn env_at(home: &Path) -> PathEnv {
        PathEnv {
            platform: crate::health::node_platform().to_string(),
            home_dir: home.to_string_lossy().into_owned(),
            env: HashMap::new(),
        }
    }

    fn source<'a>(sources: &'a SettingsSources, id: &str) -> &'a Source {
        sources
            .sources
            .iter()
            .find(|source| source.location_id == id)
            .unwrap()
    }

    /// `settings-sources-locations.test.ts`: a missing rules directory is
    /// absent, an existing empty one is present with no rules.
    #[test]
    fn a_rules_directory_is_absent_until_it_exists_then_present_even_empty() {
        let home = crate::test_support::scratch_dir("library-rules-dir");
        let absent = read_settings_sources(&env_at(&home));
        assert!(!source(&absent, "codex-permission-rules").present);
        std::fs::create_dir_all(home.join(".codex/rules")).unwrap();
        let empty = read_settings_sources(&env_at(&home));
        let rules = source(&empty, "codex-permission-rules");
        assert_eq!(
            (rules.present, rules.rules.as_deref()),
            (true, Some(&[][..])),
            "expected present with an empty rule list | received {rules:?}"
        );
    }

    /// `settings-sources.test.ts` "refuses a settings path that is a
    /// symlink": a fixed vendor path pointing elsewhere is not followed.
    #[cfg(unix)]
    #[test]
    fn a_settings_path_that_is_a_symlink_is_refused() {
        let home = crate::test_support::scratch_dir("library-settings-symlink");
        std::fs::create_dir_all(home.join(".claude")).unwrap();
        std::fs::write(home.join("elsewhere.json"), "{\"secret\": true}").unwrap();
        std::os::unix::fs::symlink(
            home.join("elsewhere.json"),
            home.join(".claude/settings.json"),
        )
        .unwrap();
        let sources = read_settings_sources(&env_at(&home));
        let settings = source(&sources, "claude-settings");
        assert_eq!(settings.failure_reason, Some(ReadFailure::NotRegularFile));
        assert_eq!(
            settings.content, None,
            "the symlink target must never be read"
        );
    }

    #[test]
    fn source_locations_follow_target_order_settings_before_hooks() {
        assert_eq!(
            settings_source_location_ids(),
            [
                "mango-settings",
                "claude-settings",
                "claude-hooks",
                "codex-settings",
                "codex-hooks",
                "codex-permission-rules",
                "cursor-settings",
            ]
        );
    }
}
